homeassistant/components/purpleair/config_flow.py

```diff
@@ -67,6 +70,43 @@ def async_get_coordinates_schema(hass: HomeAssistant) -> vol.Schema:
     )
 
 
+@callback
+def async_get_device_entity_removal_event(
+    hass: HomeAssistant, config_entry: ConfigEntry, device_id: str
+) -> asyncio.Event:
+    """Get an event that, when set, signifies all device entities have been removed."""
+    device_registry = dr.async_get(hass)
+    entity_registry = er.async_get(hass)
+
+    # Get all of the entity entries associated with the removed device:
+    device_entry = cast(dr.DeviceEntry, device_registry.async_get(device_id))
+    entity_entries = er.async_entries_for_device(
+        entity_registry, device_entry.id, include_disabled_entities=True
+    )
+
+    device_entities_removed_event = asyncio.Event()
+
+    @callback
+    def async_device_entity_state_changed(_: Event) -> None:
+        """Listen and respond when all device entities are removed."""
+        if all(
+            hass.states.get(entity_entry.entity_id) is None
+            for entity_entry in entity_entries
+        ):
+            device_entities_removed_event.set()
+
+    # Track the state change event for all entities attached to this device:
+    config_entry.async_on_unload(
```

Comment:
It probably isn't correct to cancel this listener on config entry unload— if someone adds and removes a bunch of sensors, that callback list could build up. Probably better to remove it immediately after we no longer need it (i.e., after the event is set).